lib/core/src/Ouroboros/Network/Client/Wallet.hs

```
@@ -306,51 +286,65 @@ chainSyncWithBlocks tr fromTip queue responseBuffer =
+        let points' = if null points then [Point Origin] else points
+        traceWith tr $ MsgChainFindIntersect points'
+        pure $ P.SendMsgFindIntersect points' clientStIntersect
+
+    -- Receive the result of the MsgFindIntersection request
+    clientStIntersect
+        :: P.ClientPipelinedStIntersect block (Point block) (Tip block) m Void
+    clientStIntersect = P.ClientPipelinedStIntersect
+        { P.recvMsgIntersectFound = \_point tip -> do
+            -- Here, the node tells us which  point  from the possible
+            -- intersections is the latest point on the chain.
+            -- However, we do not have to roll back to this point here;
+            -- when we send a MsgRequestNext message, the node will reply
+            -- with a MsgRollBackward message to this point first.
+            --
+            -- This behavior is not in the network specification yet, but see
+            -- https://input-output-rnd.slack.com/archives/CDA6LUXAQ/p1623322238039900
+            traceWith tr $ MsgChainTip (getTipPoint tip)
+            clientStIdle oneByOne
+        , P.recvMsgIntersectNotFound = \_tip -> do
+            -- No intersection was found.
+            -- As the read-pointer on the node could be unknown to us,
+            -- we now explicitly request the genesis point.
+            --
+            -- See also
+            -- https://input-output-rnd.slack.com/archives/CDA6LUXAQ/p1634644689103100
+            clientStNegotiateGenesis
+            }
+
+    -- Explictly negotiate the genesis point
+    clientStNegotiateGenesis
+        :: m (P.ClientPipelinedStIdle 'Z block (Point block) (Tip block) m Void)
+    clientStNegotiateGenesis = do
+        let genesis = [Point Origin]
+        traceWith tr $ MsgChainFindIntersect genesis
+        pure $ P.SendMsgFindIntersect genesis $
+            clientStIntersect
+                { P.recvMsgIntersectNotFound = \_tip ->
+                    throwIO ErrChainSyncNoIntersectGenesis
```

I guess this would be caught by `catchFromIO` in `restoreWallet` rather than causing `chainSync` to go into a retry-loop, which seems good.

---

Yes, exactly. This is an infinite loop that we can detect, so let's throw an error instead.